pkg/server/plugin/nodeattestor/gcp/iit.go

```diff
@@ -17,74 +19,80 @@ import (
 }
 
+// IITAttestorPlugin implements node attestation for agents running in GCP.
+type IITAttestorPlugin struct {
+	svidTemplate      *template.Template
+	config            *IITAttestorConfig
+	mtx               sync.Mutex
+	tokenKeyRetriever tokenKeyRetriever
+}
+
+// IITAttestorConfig is the config for IITAttestorPlugin.
 type IITAttestorConfig struct {
 	trustDomain        string
 	ProjectIDWhitelist []string `hcl:"projectid_whitelist"`
+	AgentSVIDTemplate  string   `hcl:"agent_svid_template"`
 }
 
-type IITAttestorPlugin struct {
-	tokenKeyRetriever tokenKeyRetriever
+// NewIITAttestorPlugin creates a new IITAttestorPlugin.
+func NewIITAttestorPlugin() *IITAttestorPlugin {
+	return &IITAttestorPlugin{
+		svidTemplate:      defaultAgentSVIDTemplate,
+		tokenKeyRetriever: newGooglePublicKeyRetriever(googleCertURL),
+	}
+}
 
-	mtx    sync.Mutex
-	config *IITAttestorConfig
+// templateData is the data passed to the agent SVID template.
+type templateData struct {
+	gcp.ComputeEngine
+	PluginName  string
+	TrustDomain string
 }
 
+// Attest implements the server side logic for the gcp iit node attestation plugin.
 func (p *IITAttestorPlugin) Attest(stream nodeattestor.Attest_PluginStream) error {
 	c, err := p.getConfig()
 	if err != nil {
 		return err
 	}
 
-	req, err := stream.Recv()
+	identityMetadata, err := p.validateAttestationAndExtractIdentityMetadata(stream, gcp.PluginName)
 	if err != nil {
 		return err
 	}
 
-	attestationData := req.GetAttestationData()
-	if attestationData == nil {
-		return newError("request missing attestation data")
-	}
-
-	if attestationData.Type != gcp.PluginName {
-		return newErrorf("unexpected attestation data type %q", attestationData.Type)
-	}
-
-	if req.AttestedBefore {
-		return newError("instance ID has already been attested")
-	}
-
-	identityToken := &gcp.IdentityToken{}
-	_, err = jwt.ParseWithClaims(string(req.GetAttestationData().Data), identityToken, p.tokenKeyRetriever.retrieveKey)
-	if err != nil {
-		return newErrorf("unable to parse/validate the identity token: %v", err)
-	}
-
-	if identityToken.Audience != tokenAudience {
-		return newErrorf("unexpected identity token audience %q", identityToken.Audience)
-	}
-
 	projectIDMatchesWhitelist := false
 	for _, projectID := range c.ProjectIDWhitelist {
-		if identityToken.Google.ComputeEngine.ProjectID == projectID {
+		if identityMetadata.ProjectID == projectID {
 			projectIDMatchesWhitelist = true
 			break
 		}
 	}
 	if !projectIDMatchesWhitelist {
-		return newErrorf("identity token project ID %q is not in the whitelist", identityToken.Google.ComputeEngine.ProjectID)
+		return newErrorf("identity token project ID %q is not in the whitelist", identityMetadata.ProjectID)
 	}
 
-	spiffeID := gcp.MakeSpiffeID(c.trustDomain, identityToken.Google.ComputeEngine.ProjectID, identityToken.Google.ComputeEngine.InstanceID)
+	var spiffeID bytes.Buffer
+	if err := p.svidTemplate.Execute(&spiffeID, templateData{
```

Comment:
svidTemplate is set by `Configure()` but is being not produced/consumed  while holding the lock. Although Configure() is currently only invoked once, we plan on supported config reloading at some point. The template should be added to the `IITAttestorConfig` struct (unexported so it isn't processed by the HCL decoder). This follows the pattern already established for `trustDomain`.